File: some_cipher/some_cipher.py
"""
Implementation of SomeCipher.
"""

import random

ROUNDS = 5

SBOX = [
    0xE, 0x4, 0xD, 0x1, 0x2, 0xF, 0xB, 0x8,
    0x3, 0xA, 0x6, 0xC, 0x5, 0x9, 0x0, 0x7
]

INV_SBOX = [
    0xE, 0x3, 0x4, 0x8, 0x1, 0xC, 0xA, 0xF,
    0x7, 0xD, 0x9, 0x6, 0xB, 0x2, 0x0, 0x5
]

MULTI_1 = [
    0x0, 0x1, 0x2, 0x3, 0x4, 0x5, 0x6, 0x7,
    0x8, 0x9, 0xA, 0xB, 0xC, 0xD, 0xE, 0xF
]

MULTI_4 = [
    0x0, 0x4, 0x8, 0xC, 0x3, 0x7, 0xB, 0xF,
    0x6, 0x2, 0xE, 0xA, 0x5, 0x1, 0xD, 0x9
]

MULTI_7 = [
    0x0, 0x7, 0xE, 0x9, 0xF, 0x8, 0x1, 0x6,
    0xD, 0xA, 0x3, 0x4, 0x2, 0x5, 0xC, 0xB
]

MULTI_8 = [
    0x0, 0x8, 0x3, 0xB, 0x6, 0xE, 0x5, 0xD,
    0xC, 0x4, 0xF, 0x7, 0xA, 0x2, 0x9, 0x1
]

MULTI_9 = [
    0x0, 0x9, 0x1, 0x8, 0x2, 0xB, 0x3, 0xA,
    0x4, 0xD, 0x5, 0xC, 0x6, 0xF, 0x7, 0xE
]

MULTI_12 = [
    0x0, 0xC, 0xB, 0x7, 0x5, 0x9, 0xE, 0x2,
    0xA, 0x6, 0x1, 0xD, 0xF, 0x3, 0x4, 0x8
]

# utility functions

def chunks(xs, n):
    """
    xs -> List of elements.
    n -> Length of chunk.

    Divide a list of elements into chunks of length `n`.
    """
    for i in range(0, len(xs), n):
        yield xs[i:i+n]

# procedures

def nibble_sub(ns):
    """
    ns -> List of nibbles

    Passes each nibble through a S-Box. The mapping is specified in `SBOX`.
    """
    return [SBOX[n] for n in ns]

def inv_nibble_sub(ns):
    """
    ns -> List of nibbles

    Inverse function of `nibble_sub()`.
    """
    return [INV_SBOX[n] for n in ns]

def shift_row(ns):
    """
    ns -> List of nibbles

    Shift each row of the matrix of nibbles to the right by different amounts.

    The first and second row is unchanged, while the third and fourth row are
    rotated by 1 and 2 elements to the right respectively.
    """
    assert len(ns) == 12
    n0, n1, n2, n3, n4, n5, n6, n7, n8, n9, n10, n11 = ns
    return [n0, n1, n10, n7, n4, n5, n2, n11, n8, n9, n6, n3]

def inv_shift_row(ns):
    """
    ns -> List of nibbles

    Inverse function of `shift_row()`.
    """
    assert len(ns) == 12
    n0, n1, n10, n7, n4, n5, n2, n11, n8, n9, n6, n3 = ns
    return [n0, n1, n2, n3, n4, n5, n6, n7, n8, n9, n10, n11]

def mix_column(ns):
    """
    ns -> List of nibbles

    Multiplies the matrix of nibbles by a constant matrix,

        1  1  4  9

        1  4  9  1

        4  9  1  1

        9  1  1  4
    """
    def multiply_matrix(ms):
        assert len(ms) == 4
        m0, m1, m2, m3 = ms
        return [
            MULTI_1[m0] ^ MULTI_1[m1] ^ MULTI_4[m2] ^ MULTI_9[m3],
            MULTI_1[m0] ^ MULTI_4[m1] ^ MULTI_9[m2] ^ MULTI_1[m3],
            MULTI_4[m0] ^ MULTI_9[m1] ^ MULTI_1[m2] ^ MULTI_1[m3],
            MULTI_9[m0] ^ MULTI_1[m1] ^ MULTI_1[m2] ^ MULTI_4[m3],
        ]

    assert len(ns) == 12
    return [
        n
        for ms in chunks(ns, 4)
        for n in multiply_matrix(ms)
    ]

def inv_mix_column(ns):
    """
    ns -> List of nibbles

    Inverse of `mix_column()`.

    Multiplies the matrix of nibbles by a constant matrix,

        8  12 7  7

        12 7  7  8

        7  7  8  12

        7  8  12 7
    """
    def multiply_matrix(ns):
        assert len(ns) == 4
        n0, n1, n2, n3 = ns
        return [
            MULTI_8[n0] ^ MULTI_12[n1] ^ MULTI_7[n2] ^ MULTI_7[n3],
            MULTI_12[n0] ^ MULTI_7[n1] ^ MULTI_7[n2] ^ MULTI_8[n3],
            MULTI_7[n0] ^ MULTI_7[n1] ^ MULTI_8[n2] ^ MULTI_12[n3],
            MULTI_7[n0] ^ MULTI_8[n1] ^ MULTI_12[n2] ^ MULTI_7[n3],
        ]

    assert len(ns) == 12
    return [
        n
        for ms in chunks(ns, 4)
        for n in multiply_matrix(ms)
    ]

def key_addition(ks, ns):
    """
    ks -> List of round key nibbles
    ns -> List of nibbles

    XORS each nibble with its corresponding round key nibble.
    """
    assert len(ks) == len(ns)
    return [w ^ n for (w, n) in zip(ks, ns)]

def roundf(ks, ns, i):
    """
    ks -> List of key nibbles
    ns -> List of nibbles
    i -> Round number

    Round function of SomeCipher.

    This function operates by performing the following operations in order:
    1. KeyAddition (with ith round key)
    2. NibbleSub
    3. ShiftRow
    4. MixColumn (if not the last round)

    Note that this function does not handle the addition of the final round key.
    """
    assert len(ns) == 12
    ns = shift_row(nibble_sub(ns))
    if i != ROUNDS - 1:
        ns = mix_column(ns)
    return key_addition(ks, ns)

def inv_roundf(ks, ns, i):
    """
    ks -> List of key nibbles
    ns -> List of nibbles
    i -> Round number

    Inverse round function of SomeCipher. Refer to `roundf()` for more details.
    """
    assert len(ns) == 12
    ns = key_addition(ks, ns)
    if i != ROUNDS - 1:
        ns = inv_mix_column(ns)
    return inv_nibble_sub(inv_shift_row(ns))

def encrypt_block(k, p):
    """
    k -> Key
    p -> Plaintext

    Encrypts the plaintext block with SomeCipher, given the key.
    """
    assert len(k) == len(p) == 12
    ns = key_addition(k, p)
    for i in range(0, ROUNDS):
        ns = roundf(k, ns, i)
    return ns

def decrypt_block(k, c):
    """
    k -> Key
    c -> Ciphertext

    Decrypts the ciphertext block encrypted with SomeCipher, given the key.
    """
    assert len(k) == len(c) == 12
    ns = c
    for i in range(ROUNDS - 1, -1, -1):
        ns = inv_roundf(k, ns, i)
    return key_addition(k, ns)

# main function

def main():
    k = [random.randint(0, 15) for _ in range(12)]
    p = [random.randint(0, 15) for _ in range(12)]
    print("Key: {}".format(k))
    print("Plaintext: {}".format(p))

    c = encrypt_block(k, p)
    print("Ciphertext: {}".format(c))
    d = decrypt_block(k, c)
    print("Decrypted ciphertext: {}".format(d))

    if p == d:
        print("Decrypted ciphertext is the plaintext.")
    else:
        print("ERROR: Decrypted ciphertext does not match the plaintext.")

if __name__ == "__main__":
    main()

File: some_cipher/some_cipher.cpp
#include <algorithm>
#include <iostream>

#include "some_cipher.h"

const int ROUNDS = 5;

const std::array<unsigned char, 16> SBOX = {
    0xE, 0x4, 0xD, 0x1, 0x2, 0xF, 0xB, 0x8,
    0x3, 0xA, 0x6, 0xC, 0x5, 0x9, 0x0, 0x7
};

const std::array<unsigned char, 16> INV_SBOX = {
    0xE, 0x3, 0x4, 0x8, 0x1, 0xC, 0xA, 0xF,
    0x7, 0xD, 0x9, 0x6, 0xB, 0x2, 0x0, 0x5
};

const std::array<unsigned char, 16> M1 = {
    0x0, 0x1, 0x2, 0x3, 0x4, 0x5, 0x6, 0x7,
    0x8, 0x9, 0xA, 0xB, 0xC, 0xD, 0xE, 0xF
};

const std::array<unsigned char, 16> M4 = {
    0x0, 0x4, 0x8, 0xC, 0x3, 0x7, 0xB, 0xF,
    0x6, 0x2, 0xE, 0xA, 0x5, 0x1, 0xD, 0x9
};

const std::array<unsigned char, 16> M7 = {
    0x0, 0x7, 0xE, 0x9, 0xF, 0x8, 0x1, 0x6,
    0xD, 0xA, 0x3, 0x4, 0x2, 0x5, 0xC, 0xB
};

const std::array<unsigned char, 16> M8 = {
    0x0, 0x8, 0x3, 0xB, 0x6, 0xE, 0x5, 0xD,
    0xC, 0x4, 0xF, 0x7, 0xA, 0x2, 0x9, 0x1
};

const std::array<unsigned char, 16> M9 = {
    0x0, 0x9, 0x1, 0x8, 0x2, 0xB, 0x3, 0xA,
    0x4, 0xD, 0x5, 0xC, 0x6, 0xF, 0x7, 0xE
};

const std::array<unsigned char, 16> M12 = {
    0x0, 0xC, 0xB, 0x7, 0x5, 0x9, 0xE, 0x2,
    0xA, 0x6, 0x1, 0xD, 0xF, 0x3, 0x4, 0x8
};

diffs differences(const nibs ns0, const nibs ns1) {
    diffs ds;
    for (int i = 0; i < 12; ++i) ds[i] = ns0[i] != ns1[i];
    return ds;
}

nibs nibble_sub(const nibs ns) {
    auto os = ns;
    for (int i = 0; i < 12; ++i) os[i] = SBOX[ns[i]];
    return os;
}

nibs inv_nibble_sub(const nibs ns) {
    auto os = ns;
    for (int i = 0; i < 12; ++i) os[i] = INV_SBOX[ns[i]];
    return os;
}

nibs shift_row(const nibs ns) {
    auto os = ns;
    std::swap(os[2], os[6]);
    std::swap(os[2], os[10]);
    std::swap(os[3], os[11]);
    std::swap(os[3], os[7]);
    return os;
}

nibs inv_shift_row(const nibs ns) {
    auto os = ns;
    std::swap(os[2], os[6]);
    std::swap(os[6], os[10]);
    std::swap(os[3], os[7]);
    std::swap(os[3], os[11]);
    return os;
}

nibs mix_column(const nibs ns) {
    nibs os;
    os[0] = M1[ns[0]] ^ M1[ns[1]] ^ M4[ns[2]] ^ M9[ns[3]];
    os[1] = M1[ns[0]] ^ M4[ns[1]] ^ M9[ns[2]] ^ M1[ns[3]];
    os[2] = M4[ns[0]] ^ M9[ns[1]] ^ M1[ns[2]] ^ M1[ns[3]];
    os[3] = M9[ns[0]] ^ M1[ns[1]] ^ M1[ns[2]] ^ M4[ns[3]];

    os[4] = M1[ns[4]] ^ M1[ns[5]] ^ M4[ns[6]] ^ M9[ns[7]];
    os[5] = M1[ns[4]] ^ M4[ns[5]] ^ M9[ns[6]] ^ M1[ns[7]];
    os[6] = M4[ns[4]] ^ M9[ns[5]] ^ M1[ns[6]] ^ M1[ns[7]];
    os[7] = M9[ns[4]] ^ M1[ns[5]] ^ M1[ns[6]] ^ M4[ns[7]];

    os[8] = M1[ns[8]] ^ M1[ns[9]] ^ M4[ns[10]] ^ M9[ns[11]];
    os[9] = M1[ns[8]] ^ M4[ns[9]] ^ M9[ns[10]] ^ M1[ns[11]];
    os[10] = M4[ns[8]] ^ M9[ns[9]] ^ M1[ns[10]] ^ M1[ns[11]];
    os[11] = M9[ns[8]] ^ M1[ns[9]] ^ M1[ns[10]] ^ M4[ns[11]];
    return os;
}

nibs inv_mix_column(const nibs ns) {
    nibs os;
    os[0] = M8[ns[0]] ^ M12[ns[1]] ^ M7[ns[2]] ^ M7[ns[3]];
    os[1] = M12[ns[0]] ^ M7[ns[1]] ^ M7[ns[2]] ^ M8[ns[3]];
    os[2] = M7[ns[0]] ^ M7[ns[1]] ^ M8[ns[2]] ^ M12[ns[3]];
    os[3] = M7[ns[0]] ^ M8[ns[1]] ^ M12[ns[2]] ^ M7[ns[3]];

    os[4] = M8[ns[4]] ^ M12[ns[5]] ^ M7[ns[6]] ^ M7[ns[7]];
    os[5] = M12[ns[4]] ^ M7[ns[5]] ^ M7[ns[6]] ^ M8[ns[7]];
    os[6] = M7[ns[4]] ^ M7[ns[5]] ^ M8[ns[6]] ^ M12[ns[7]];
    os[7] = M7[ns[4]] ^ M8[ns[5]] ^ M12[ns[6]] ^ M7[ns[7]];

    os[8] = M8[ns[8]] ^ M12[ns[9]] ^ M7[ns[10]] ^ M7[ns[11]];
    os[9] = M12[ns[8]] ^ M7[ns[9]] ^ M7[ns[10]] ^ M8[ns[11]];
    os[10] = M7[ns[8]] ^ M7[ns[9]] ^ M8[ns[10]] ^ M12[ns[11]];
    os[11] = M7[ns[8]] ^ M8[ns[9]] ^ M12[ns[10]] ^ M7[ns[11]];
    return os;
}

nibs key_addition(const nibs ks, const nibs ns) {
    nibs os;
    for (int i = 0; i < 12; ++i) os[i] = ks[i] ^ ns[i];
    return os;
}

nibs roundf(const nibs ks, const nibs ns, const int i) {
    nibs os = shift_row(nibble_sub(ns));
    if (i != ROUNDS - 1) os = mix_column(os);
    return key_addition(ks, os);
}

nibs inv_roundf(const nibs ks, const nibs ns, const int i) {
    nibs os = key_addition(ks, ns);
    if (i != ROUNDS - 1) os = inv_mix_column(os);
    return inv_nibble_sub(inv_shift_row(os));
}

nibs encrypt_block(const nibs ks, const nibs ps) {
    nibs ns = key_addition(ks, ps);
    for (int i = 0; i < ROUNDS; ++i) ns = roundf(ks, ns, i);
    return ns;
}

nibs decrypt_block(const nibs ks, const nibs cs) {
    nibs ns = cs;
    for (int i = ROUNDS - 1; i >= 0; --i) ns = inv_roundf(ks, ns, i);
    return key_addition(ks, ns);
}
